hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/ozoneimpl/ContainerScanHelper.java

```diff
@@ -0,0 +1,154 @@
+      log.debug("Container [{}] has been deleted during the data scan.", containerId);
+    } else {
+      try {
+        controller.updateContainerChecksum(containerId, result.getDataTree());
+      } catch (IOException ex) {
+        log.warn("Failed to update container checksum after scan of container {}", containerId, ex);
+      }
+      if (!result.isHealthy()) {
+        handleUnhealthyScanResult(containerId, result);
+      }
+      metrics.incNumContainersScanned();
+    }
+
+    Instant now = Instant.now();
+    if (!result.isDeleted()) {
+      controller.updateDataScanTimestamp(containerId, now);
+    }
+    // Even if the container was deleted, mark the scan as completed since we already logged it as starting.
+    logScanCompleted(containerData, now);
+  }
+
+  public void handleUnhealthyScanResult(long containerID, ScanResult result) throws IOException {
+
+    log.error("Corruption detected in container [{}]. Marking it UNHEALTHY. {}", containerID, result);
+    if (log.isDebugEnabled()) {
+      StringBuilder allErrorString = new StringBuilder();
+      result.getErrors().forEach(r -> allErrorString.append(r).append('\n'));
+      log.debug("Complete list of errors detected while scanning container {}:\n{}", containerID, allErrorString);
+    }
+
+    // Only increment the number of unhealthy containers if the container was not already unhealthy.
+    // TODO HDDS-11593: Scanner counters will start from zero
+    //  at the beginning of each run, so this will need to be incremented for every unhealthy container seen
+    //  regardless of its previous state.
+    boolean containerMarkedUnhealthy = controller.markContainerUnhealthy(containerID, result);
+    if (containerMarkedUnhealthy) {
+      metrics.incNumUnHealthyContainers();
+    }
+  }
+
+  public boolean shouldScanMetadata(Container<?> container) {
+    if (container == null) {
+      return false;
+    }
+    long containerID = container.getContainerData().getContainerID();
+
+    HddsVolume containerVolume = container.getContainerData().getVolume();
+    if (containerVolume.isFailed()) {
+      log.debug("Skipping scan for container {} since its volume {} has failed.", containerID, containerVolume);
+      return false;
+    }
+
+    return !recentlyScanned(container.getContainerData());
+  }
+
+  public boolean shouldScanData(Container<?> container) {
+    return shouldScanMetadata(container) && container.shouldScanData();
+  }
+
+  private boolean recentlyScanned(ContainerData containerData) {
+    Optional<Instant> lastScanTime = containerData.lastDataScanTime();
+    Instant now = Instant.now();
+    // Container is considered recently scanned if it was scanned within the
+    // configured time frame. If the optional is empty, the container was
+    // never scanned.
+    boolean recentlyScanned = lastScanTime.map(scanInstant ->
+            Duration.between(now, scanInstant).abs()
+                .compareTo(Duration.ofMillis(minScanGap)) < 0)
```

Comment:
This is specific to `minScanGap` for On-demand scanner. Right now, the default is 15m. Which means we have to wait for the on-demand scanner to run on the unhealthy container and wait 15 mins before triggering reconciliation as reconciliation also triggers an on-demand scan at the end, That might get dropped because of this.